authentication/oidc.go

```diff
@@ -45,71 +47,127 @@ type OIDCConfig struct {
-		Name:      "tenant_registration_failures",
-		Help:      "The number of tenants for which OIDC provider instantiation failed.",
+		Name:      "tenants_failing_registrations",
+		Help:      "The number of tenants for which OIDC provider instantiation is failing.",
 	})
 
-	reg.MustRegister(tenantErrors)
+	reg.MustRegister(tenantsFailing)
 
-	for _, c := range configs {
-		p, err := NewProvider(context.TODO(), logger, getCookieForTenant(c.Tenant), "/"+c.Tenant, c.OIDCConfig)
-		if err != nil {
-			warnings = append(warnings, fmt.Errorf("failed to instantiate OIDC provider for tenant %q: %w", c.Tenant, err))
+	return &OIDCHandlers{
+		handlers:    make(map[string]http.Handler),
+		middlewares: make(map[string]Middleware),
+		logger:      l,
+		retryGauge:  tenantsFailing,
+	}
+}
 
-			tenantErrors.Inc()
+// AddOIDCForTenant adds a new handler and middleware for a new tenant, based on
+// the tenant OIDC configuration. If a connection with the OIDC provider
+// cannot be established, the method will retry to connect with an increasing
+// backoff.
+func (oh *OIDCHandlers) AddOIDCForTenant(prefix string, config TenantOIDCConfig) {
+	const (
+		loginRoute    = "/login"
+		callbackRoute = "/callback"
+	)
+
+	ctx := context.Background()
+	b := backoff.New(ctx, backoff.Config{
+		Min:        500 * time.Millisecond,
+		Max:        5 * time.Second,
+		MaxRetries: 0, // retry indefinitely
+	})
 
-			continue
-		}
+	go func() {
+		var isFailing bool
+		for b.Reset(); b.Ongoing(); {
+			p, err := NewProvider(ctx, oh.logger, getCookieForTenant(config.Tenant), "/"+config.Tenant, config.OIDCConfig)
+			if err != nil {
+				level.Warn(oh.logger).Log("msg", "failed to instantiate OIDC provider for tenant", "tenant", config.Tenant, "error", err)
 
-		r := chi.NewRouter()
+				if !isFailing {
+					oh.retryGauge.Inc()
+					isFailing = true
+				}
+
+				b.Wait()
+				continue
+			}
 
-		const (
-			loginRoute    = "/login"
-			callbackRoute = "/callback"
-		)
+			r := chi.NewRouter()
+			r.Handle(loginRoute, otelhttp.WithRouteTag(prefix+loginRoute, p.LoginHandler()))
+			r.Handle(callbackRoute, otelhttp.WithRouteTag(prefix+callbackRoute, p.CallbackHandler()))
 
-		r.Handle(loginRoute, otelhttp.WithRouteTag(prefix+loginRoute, p.LoginHandler()))
-		r.Handle(callbackRoute, otelhttp.WithRouteTag(prefix+callbackRoute, p.CallbackHandler()))
+			oh.mtx.Lock()
+			oh.handlers[config.Tenant] = r
+			oh.middlewares[config.Tenant] = p.Middleware()
+			oh.mtx.Unlock()
 
-		handlers[c.Tenant] = r
-		middlewares[c.Tenant] = p.Middleware()
+			level.Info(oh.logger).Log("msg", "OIDC provider instantiated for tenant", "tenant", config.Tenant)
+
+			if isFailing {
+				oh.retryGauge.Dec()
+			}
+
+			return
+		}
+	}()
+}
+
+// GetTenantMiddleware an OIDC middleware for the specified tenant, if found.
+func (oh *OIDCHandlers) GetTenantMiddleware(tenant string) (Middleware, bool) {
+	oh.mtx.RLock()
+	mw, ok := oh.middlewares[tenant]
+	oh.mtx.RUnlock()
+
+	if !ok {
```

Comment:
We can simplify these lässt lines and always do `return mw, ok` no?

Reply:
Totally :sweat_smile: